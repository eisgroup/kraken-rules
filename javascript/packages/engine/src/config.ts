/*
 *  Copyright 2020 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { debug } from './debugger/Debugger'

/**
 * Configures Kraken evaluation. It can be configured from the console.
 * Open the console and  modify the config options
 * @example
 * global.Kraken.debug = false;
 * global.Kraken.info = false;
 * global.Kraken.debug = false;
 * @see Logger
 * @since 1.1.1
 */
export interface KrakenConfig {
    logger: {
        enabled: boolean
        debug: boolean
        info: boolean
        error: boolean
        warning: boolean
        entriesCount: number
        logs: string[]
        getLogs(): void
        clear(): void
    }
    debugger: debug.api.Debugger
}
export type WithKraken = {
    Kraken: KrakenConfig
}

function init(): void {
    const logger: KrakenConfig['logger'] = {
        enabled: process.env.NODE_ENV !== 'production',
        debug: false,
        info: true,
        error: true,
        warning: true,
        entriesCount: 2000,
        logs: [],
        clear(): void {
            const config = (global as unknown as WithKraken).Kraken
            if (config.logger.logs) {
                config.logger.logs = []
            }
        },
        getLogs(): void {
            const config = (global as unknown as WithKraken).Kraken
            // applicable only for the browser
            if (config.logger.logs && document && window) {
                const blob = new Blob([config.logger.logs.join('\n')], { type: 'text/plain' })
                const url = window.URL.createObjectURL(blob)
                const a = document.createElement('a')
                a.style.display = 'none'
                a.href = url
                a.download = `kraken-logs-${new Date().getTime()}.txt`
                document.body.appendChild(a)
                a.click()
                window.URL.revokeObjectURL(url)
                document.body.removeChild(a)
            }
        },
    }
    if (global) {
        ;(global as unknown as WithKraken).Kraken = {
            logger,
            debugger: new debug.impl.DevToolsDebugger(),
        }
        return
    }
}

function getConfig(): KrakenConfig {
    if (global) {
        return (global as unknown as { Kraken: KrakenConfig })['Kraken']
    }
    throw new Error('Kraken configuration was not initialized')
}

export const krakenConfig = { init, getConfig }
