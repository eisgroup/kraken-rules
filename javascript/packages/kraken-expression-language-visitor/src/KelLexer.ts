import { Common } from './antlr/generated/Common'

export class KelLexer extends Common {
    /** @override */
    popMode(): number {
        return this._modeStack.isEmpty ? Common.DEFAULT_MODE : super.popMode()
    }
}
