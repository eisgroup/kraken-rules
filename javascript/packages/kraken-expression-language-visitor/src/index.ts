export { createParser } from "./parse";
export { KelVisitor } from "./antlr/generated/KelVisitor";
export { SyntaxError } from "./KrakenParser";
export * from "./antlr/generated/Kel";

export { ParserRuleContext } from "antlr4ts";
export { AbstractParseTreeVisitor } from "antlr4ts/tree/AbstractParseTreeVisitor";
export { ErrorNode } from "antlr4ts/tree/ErrorNode";
export { Interval } from "antlr4ts/misc/Interval";
