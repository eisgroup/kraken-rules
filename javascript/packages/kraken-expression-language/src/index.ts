export * from './scope/Scope'
export * from './scope/ScopeType'
export * from './visitor/Node'
export * from './visitor/AstGeneratingVisitor'
export * from './visitor/NodeUtils'

export { KelParser, SyntaxError } from 'kraken-expression-language-visitor'

export * from './visitor/autocomplete/CompletionItem'
export * from './visitor/type-at-location/LocationInfo'

export * from './public/Cache'
export * from './public/KelService'
export * from './public/ScopeDeserializer'

export { Type } from './type/Types'
