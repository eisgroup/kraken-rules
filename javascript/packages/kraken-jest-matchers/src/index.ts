/*
 *  Copyright 2022 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { matchers as dateTimeMatchers } from './dateTimeMatchers'
import { matchers as ruleEvalResultMatchers } from './ruleEvalResultMatchers'
import { matchers as entryPointResultMatchers } from './entrypointResultMatchers'

export const matchers = {
    ...dateTimeMatchers,
    ...ruleEvalResultMatchers,
    ...entryPointResultMatchers,
}

export { ExpectedResult } from './entrypointResultMatchers'
export { MatchableResults } from './entrypointResultMatchers'

expect.extend(matchers)
