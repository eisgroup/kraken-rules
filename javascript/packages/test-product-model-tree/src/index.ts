export * from './generated/kraken_model_tree_Policy'
export * from './generated/kraken_model_tree_PolicyExtended'