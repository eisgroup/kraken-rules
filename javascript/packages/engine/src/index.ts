/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
import { krakenConfig } from './config'
krakenConfig.init()

// spi
export * from './engine/contexts/info/DataObjectInfoResolver'
export * from './engine/contexts/info/ContextInstanceInfoResolver'

// engine
export * from './engine/executer/SyncEngine'

export * from './bundle-cache/EntryPointBundleCache'
export * from './bundle-cache/delta-cache/RepoClientCache'
export * from './bundle-cache/delta-cache/DeltaBundleCache'
export * from './bundle-cache/dimension-set-cache/DimensionSetBundleCache'
export * from './bundle-cache/dimension-set-cache/DimensionSetCacheLogger'
export * from './bundle-cache/expression-context-manager/ExpressionContextManager'
export * from './bundle-cache/expression-context-manager/ExpressionContextManagerImpl'

export { registry, KelFunction, FunctionParameter } from './engine/runtime/expressions/ExpressionEvaluator'
export { FunctionScope } from './engine/runtime/expressions/functionLibrary/Registry'
export * from './engine/runtime/EvaluationMode'

// Payload result utils
export * from './engine/results/PayloadResultCreator'
export * from './engine/results/PayloadResultTypeChecker'
export { ValueChangedEvent } from './engine/results/ValueChangedEvent'

// Condition result utils
export { conditionEvaluationTypeChecker } from './dto/DefaultConditionEvaluationResult'

export { DataContext } from './engine/contexts/data/DataContext'

// field metadata reducer
export * from './engine/results/field_metadata_reducer/FieldErrorMessage'

export * from './models/EntryPointBundle'
export * from './models/ContextModelTree'

// reducer
export * from './engine/results/RuleOverrideContextExtractor'
export * from './engine/results/Reducer'
export * from './engine/results/Localization'

// field metadata reducer
export * from './engine/results/field_metadata_reducer/FieldMetadataReducer'

// error
export * from './error/KrakenRuntimeError'

// re-export
export {
    ContextInstanceInfo,
    RuleOverride,
    RuleInfo,
    ContextFieldInfo,
    EntryPointResult,
    ConditionEvaluationResult,
    FieldEvaluationResult,
    RuleEvaluationResults,
    PayloadResultType,
    PayloadResult,
    ErrorAwarePayloadResult,
    ErrorMessage,
    ValidationPayloadResult,
    AccessibilityPayloadResult,
    VisibilityPayloadResult,
    DefaultValuePayloadResult,
    AssertionPayloadResult,
    SizePayloadResult,
    SizeRangePayloadResult,
    LengthPayloadResult,
    RegExpPayloadResult,
    UsagePayloadResult,
    ValueListPayloadResult,
    RuleEvent,
    FieldMetadata,
    FieldMetadataResult,
    ConditionEvaluation,
} from 'kraken-engine-api'
