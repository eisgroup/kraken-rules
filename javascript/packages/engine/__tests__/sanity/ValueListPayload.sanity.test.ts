/*
 *  Copyright 2023 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { sanityMocks } from './_AutoPolicyObject.mocks'
import { sanityEngine } from './_SanityEngine'
import { FieldMetadataReducer } from '../../src'

describe('Engine Sanity ValueList Payload Test', () => {
    const { valueListValid, valueListInvalid } = sanityMocks
    it("should execute 'ValueListPayload' entrypoint with valid data", () => {
        const results = sanityEngine.evaluate(valueListValid(), 'ValueListPayload')
        const fieldResults = new FieldMetadataReducer().reduce(results)

        expect(results).k_toMatchResultsStats({ total: 4, critical: 0 })
        expect(results).k_toHaveExpressionsFailures(0)
        expect({ entryPointResults: results, fieldResults }).k_toMatchResultsSnapshots()
    })
    it("should execute 'ValueListPayload' entrypoint with invalid data", () => {
        const results = sanityEngine.evaluate(valueListInvalid(), 'ValueListPayload')
        const fieldResults = new FieldMetadataReducer().reduce(results)

        expect(results).k_toMatchResultsStats({ total: 4, critical: 4 })
        expect(results).k_toHaveExpressionsFailures(0)
        expect({ entryPointResults: results, fieldResults }).k_toMatchResultsSnapshots()
    })
})
