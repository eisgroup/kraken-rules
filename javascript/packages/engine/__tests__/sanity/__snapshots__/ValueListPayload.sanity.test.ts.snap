// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Engine Sanity ValueList Payload Test should execute 'ValueListPayload' entrypoint with invalid data 1`] = `
Object {
  "entryPointResults": Object {
    "evaluationTimestamp": Any<Date>,
    "results": Object {
      "Policy:12:createdFromPolicyRev": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "12",
          "contextName": "Policy",
          "fieldName": "createdFromPolicyRev",
          "fieldPath": "createdFromPolicyRev",
          "fieldType": "INTEGER",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": Object {
                "errorCode": "code",
                "errorMessage": "err-integer-value-list",
                "templateVariables": Array [],
              },
              "success": false,
              "type": 2048,
              "validationSeverity": "critical",
              "valueList": Object {
                "valueType": "DECIMAL",
                "values": Array [
                  1,
                  2,
                  3,
                  4,
                  5,
                  6,
                  7,
                  8,
                ],
              },
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "VALUE_LIST",
              "ruleName": "Policy revision must be within allowed values",
              "targetPath": "createdFromPolicyRev",
            },
          },
        ],
      },
      "Policy:12:policyCurrency": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "12",
          "contextName": "Policy",
          "fieldName": "policyCurrency",
          "fieldPath": "policyCurrency",
          "fieldType": "STRING",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": Object {
                "errorCode": "code",
                "errorMessage": "err-string-value-list",
                "templateVariables": Array [],
              },
              "success": false,
              "type": 2048,
              "validationSeverity": "critical",
              "valueList": Object {
                "valueType": "STRING",
                "values": Array [
                  "USD",
                  "CAD",
                ],
              },
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "VALUE_LIST",
              "ruleName": "Policy currency must be within allowed values",
              "targetPath": "policyCurrency",
            },
          },
        ],
      },
      "Policy:12:policyValue": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "12",
          "contextName": "Policy",
          "fieldName": "policyValue",
          "fieldPath": "policyValue",
          "fieldType": "MONEY",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": undefined,
              "success": false,
              "type": 2048,
              "validationSeverity": "critical",
              "valueList": Object {
                "valueType": "DECIMAL",
                "values": Array [
                  100,
                  200,
                  350,
                  400.99,
                ],
              },
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "VALUE_LIST",
              "ruleName": "Policy value must be within allowed values",
              "targetPath": "policyValue",
            },
          },
        ],
      },
      "RRCoverage:21:limitAmount": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "21",
          "contextName": "RRCoverage",
          "fieldName": "limitAmount",
          "fieldPath": "limitAmount",
          "fieldType": "DECIMAL",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": undefined,
              "success": false,
              "type": 2048,
              "validationSeverity": "critical",
              "valueList": Object {
                "valueType": "DECIMAL",
                "values": Array [
                  800000,
                  900000,
                  1000000,
                  800000.99,
                ],
              },
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "RRCoverage",
              "payloadtype": "VALUE_LIST",
              "ruleName": "RRCoverage limit must be within allowed values",
              "targetPath": "limitAmount",
            },
          },
        ],
      },
    },
  },
  "fieldResults": Object {
    "Policy:12:createdFromPolicyRev": Object {
      "fieldType": "INTEGER",
      "id": "Policy:12:createdFromPolicyRev",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "createdFromPolicyRev",
      "ruleResults": Array [
        Object {
          "errorCode": "code",
          "errorMessage": "err-integer-value-list",
          "isFailed": true,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": Object {
              "errorCode": "code",
              "errorMessage": "err-integer-value-list",
              "templateVariables": Array [],
            },
            "success": false,
            "type": 2048,
            "validationSeverity": "critical",
            "valueList": Object {
              "valueType": "DECIMAL",
              "values": Array [
                1,
                2,
                3,
                4,
                5,
                6,
                7,
                8,
              ],
            },
          },
          "ruleName": "Policy revision must be within allowed values",
          "severity": "critical",
          "templateVariables": Array [],
        },
      ],
    },
    "Policy:12:policyCurrency": Object {
      "fieldType": "STRING",
      "id": "Policy:12:policyCurrency",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "policyCurrency",
      "ruleResults": Array [
        Object {
          "errorCode": "code",
          "errorMessage": "err-string-value-list",
          "isFailed": true,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": Object {
              "errorCode": "code",
              "errorMessage": "err-string-value-list",
              "templateVariables": Array [],
            },
            "success": false,
            "type": 2048,
            "validationSeverity": "critical",
            "valueList": Object {
              "valueType": "STRING",
              "values": Array [
                "USD",
                "CAD",
              ],
            },
          },
          "ruleName": "Policy currency must be within allowed values",
          "severity": "critical",
          "templateVariables": Array [],
        },
      ],
    },
    "Policy:12:policyValue": Object {
      "fieldType": "MONEY",
      "id": "Policy:12:policyValue",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "policyValue",
      "ruleResults": Array [
        Object {
          "errorCode": "rule-value-list-error",
          "errorMessage": "Value must be one of: {{0}}",
          "isFailed": true,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": undefined,
            "success": false,
            "type": 2048,
            "validationSeverity": "critical",
            "valueList": Object {
              "valueType": "DECIMAL",
              "values": Array [
                100,
                200,
                350,
                400.99,
              ],
            },
          },
          "ruleName": "Policy value must be within allowed values",
          "severity": "critical",
          "templateVariables": Array [
            "100, 200, 350, 400.99",
          ],
        },
      ],
    },
    "RRCoverage:21:limitAmount": Object {
      "fieldType": "DECIMAL",
      "id": "RRCoverage:21:limitAmount",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "limitAmount",
      "ruleResults": Array [
        Object {
          "errorCode": "rule-value-list-error",
          "errorMessage": "Value must be one of: {{0}}",
          "isFailed": true,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": undefined,
            "success": false,
            "type": 2048,
            "validationSeverity": "critical",
            "valueList": Object {
              "valueType": "DECIMAL",
              "values": Array [
                800000,
                900000,
                1000000,
                800000.99,
              ],
            },
          },
          "ruleName": "RRCoverage limit must be within allowed values",
          "severity": "critical",
          "templateVariables": Array [
            "800000, 900000, 1000000, 800000.99",
          ],
        },
      ],
    },
  },
}
`;

exports[`Engine Sanity ValueList Payload Test should execute 'ValueListPayload' entrypoint with valid data 1`] = `
Object {
  "entryPointResults": Object {
    "evaluationTimestamp": Any<Date>,
    "results": Object {
      "Policy:12:createdFromPolicyRev": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "12",
          "contextName": "Policy",
          "fieldName": "createdFromPolicyRev",
          "fieldPath": "createdFromPolicyRev",
          "fieldType": "INTEGER",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": Object {
                "errorCode": "code",
                "errorMessage": "err-integer-value-list",
                "templateVariables": Array [],
              },
              "success": true,
              "type": 2048,
              "validationSeverity": "critical",
              "valueList": Object {
                "valueType": "DECIMAL",
                "values": Array [
                  1,
                  2,
                  3,
                  4,
                  5,
                  6,
                  7,
                  8,
                ],
              },
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "VALUE_LIST",
              "ruleName": "Policy revision must be within allowed values",
              "targetPath": "createdFromPolicyRev",
            },
          },
        ],
      },
      "Policy:12:policyCurrency": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "12",
          "contextName": "Policy",
          "fieldName": "policyCurrency",
          "fieldPath": "policyCurrency",
          "fieldType": "STRING",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": Object {
                "errorCode": "code",
                "errorMessage": "err-string-value-list",
                "templateVariables": Array [],
              },
              "success": true,
              "type": 2048,
              "validationSeverity": "critical",
              "valueList": Object {
                "valueType": "STRING",
                "values": Array [
                  "USD",
                  "CAD",
                ],
              },
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "VALUE_LIST",
              "ruleName": "Policy currency must be within allowed values",
              "targetPath": "policyCurrency",
            },
          },
        ],
      },
      "Policy:12:policyValue": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "12",
          "contextName": "Policy",
          "fieldName": "policyValue",
          "fieldPath": "policyValue",
          "fieldType": "MONEY",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": undefined,
              "success": true,
              "type": 2048,
              "validationSeverity": "critical",
              "valueList": Object {
                "valueType": "DECIMAL",
                "values": Array [
                  100,
                  200,
                  350,
                  400.99,
                ],
              },
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "VALUE_LIST",
              "ruleName": "Policy value must be within allowed values",
              "targetPath": "policyValue",
            },
          },
        ],
      },
      "RRCoverage:21:limitAmount": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "21",
          "contextName": "RRCoverage",
          "fieldName": "limitAmount",
          "fieldPath": "limitAmount",
          "fieldType": "DECIMAL",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 2,
            "overrideInfo": Object {
              "overridable": false,
              "overrideApplicable": false,
              "overrideGroup": undefined,
            },
            "payloadResult": Object {
              "message": undefined,
              "success": true,
              "type": 2048,
              "validationSeverity": "critical",
              "valueList": Object {
                "valueType": "DECIMAL",
                "values": Array [
                  800000,
                  900000,
                  1000000,
                  800000.99,
                ],
              },
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "RRCoverage",
              "payloadtype": "VALUE_LIST",
              "ruleName": "RRCoverage limit must be within allowed values",
              "targetPath": "limitAmount",
            },
          },
        ],
      },
    },
  },
  "fieldResults": Object {
    "Policy:12:createdFromPolicyRev": Object {
      "fieldType": "INTEGER",
      "id": "Policy:12:createdFromPolicyRev",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "createdFromPolicyRev",
      "ruleResults": Array [
        Object {
          "errorCode": "code",
          "errorMessage": "err-integer-value-list",
          "isFailed": false,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": Object {
              "errorCode": "code",
              "errorMessage": "err-integer-value-list",
              "templateVariables": Array [],
            },
            "success": true,
            "type": 2048,
            "validationSeverity": "critical",
            "valueList": Object {
              "valueType": "DECIMAL",
              "values": Array [
                1,
                2,
                3,
                4,
                5,
                6,
                7,
                8,
              ],
            },
          },
          "ruleName": "Policy revision must be within allowed values",
          "severity": "critical",
          "templateVariables": Array [],
        },
      ],
    },
    "Policy:12:policyCurrency": Object {
      "fieldType": "STRING",
      "id": "Policy:12:policyCurrency",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "policyCurrency",
      "ruleResults": Array [
        Object {
          "errorCode": "code",
          "errorMessage": "err-string-value-list",
          "isFailed": false,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": Object {
              "errorCode": "code",
              "errorMessage": "err-string-value-list",
              "templateVariables": Array [],
            },
            "success": true,
            "type": 2048,
            "validationSeverity": "critical",
            "valueList": Object {
              "valueType": "STRING",
              "values": Array [
                "USD",
                "CAD",
              ],
            },
          },
          "ruleName": "Policy currency must be within allowed values",
          "severity": "critical",
          "templateVariables": Array [],
        },
      ],
    },
    "Policy:12:policyValue": Object {
      "fieldType": "MONEY",
      "id": "Policy:12:policyValue",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "policyValue",
      "ruleResults": Array [
        Object {
          "errorCode": "rule-value-list-error",
          "errorMessage": "Value must be one of: {{0}}",
          "isFailed": false,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": undefined,
            "success": true,
            "type": 2048,
            "validationSeverity": "critical",
            "valueList": Object {
              "valueType": "DECIMAL",
              "values": Array [
                100,
                200,
                350,
                400.99,
              ],
            },
          },
          "ruleName": "Policy value must be within allowed values",
          "severity": "critical",
          "templateVariables": Array [
            "100, 200, 350, 400.99",
          ],
        },
      ],
    },
    "RRCoverage:21:limitAmount": Object {
      "fieldType": "DECIMAL",
      "id": "RRCoverage:21:limitAmount",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "limitAmount",
      "ruleResults": Array [
        Object {
          "errorCode": "rule-value-list-error",
          "errorMessage": "Value must be one of: {{0}}",
          "isFailed": false,
          "isOverridable": false,
          "isOverridden": false,
          "payloadResult": Object {
            "message": undefined,
            "success": true,
            "type": 2048,
            "validationSeverity": "critical",
            "valueList": Object {
              "valueType": "DECIMAL",
              "values": Array [
                800000,
                900000,
                1000000,
                800000.99,
              ],
            },
          },
          "ruleName": "RRCoverage limit must be within allowed values",
          "severity": "critical",
          "templateVariables": Array [
            "800000, 900000, 1000000, 800000.99",
          ],
        },
      ],
    },
  },
}
`;
